### PR TITLE
Add inferred runtime identifiers

### PR DESCRIPTION
**Pull request by ericstj**
During source build it's desirable to add RIDs to the runtime graph
without actually. Making a change to sources.

For example: the tar-ball is produced and then built on some new distro.

The added RID should follow the compatibility rules for existing RIDs
so we parse it and try to find the most appropriate RuntimeGroup to add
it to.

cc @tmds @omajid

### CODE REVIEW

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/RID.cs [hunk 1]
```diff
@@ -40,6 +47,120 @@ public override string ToString()
...
+
+            // qualifier is indistinguishable from arch so we cannot distinguish it for parsing purposes
+            Debug.Assert(ArchitectureDelimiter == QualifierDelimiter);
+
+            for (int i = 0; i < runtimeIdentifier.Length; i++)
+            {
+                char current = runtimeIdentifier[i];
+                partLength = i - partStart;
+
+                switch (parseState)
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/src/RID.cs:76]: Did you consider using a regular expression for parsing instead?

**Comment by ViktorHofer** [comment]
Does anyone except dotnet/runtime mutate the runtime graph? Would dotnet/runtime be a better place for the runtime graph code to live?

**Comment by ericstj** [comment]
> Does anyone except dotnet/runtime mutate the runtime graph? Would dotnet/runtime be a better place for the runtime graph code to live?

This has a single consumer in dotnet\runtime: The platforms package.  If there was a good place in dotnet/runtime to both build and test tasks we could move it, though I'm not seeing a ton of value in doing so..  At the time this was [written](https://github.com/dotnet/buildtools/pull/1689) that wasn't the case.  I don't think we should block the addition of this feature on moving this out of arcade.  Perhaps when we want to remove the pkgproj stuff we can move this to dotnet/runtime.

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/RID.cs (hunk 1 above)
Comment (ericstj) [reply src/Microsoft.DotNet.Build.Tasks.Packaging/src/RID.cs:76]: Yeah I had a regex and it was much more difficult to read once it was complicated enough to handle all cases and capture all fields that we needed.  This simple state machine is easier to read and debug (and happens to be more efficient though that's not a real concern for this one shot task).

**Comment by ericstj** [comment]
Note to reviewers: best to review all changes after first refactoring commit (checkout GitHub's shift-click feature to reduce the commit range).

**Comment by ViktorHofer** [comment]
> This has a single consumer in dotnet\runtime: The platforms package. If there was a good place in dotnet/runtime to both build and test tasks we could move it, though I'm not seeing a ton of value in doing so.. At the time this was written that wasn't the case. I don't think we should block the addition of this feature on moving this out of arcade. Perhaps when we want to remove the pkgproj stuff we can move this to dotnet/runtime.

Unrelated to this PR but let me respond. If there's only a single consumer then the code shouldn't live in Arcade. There is infrastructure in dotnet/runtime that allows to build tasks, test and consume them.

> I don't think we should block the addition of this feature on moving this out of arcade.

Neither do I. It was a side comment unrelated to this PR as I wasn't aware that this code exists in the packaging code.

**Review by ViktorHofer — APPROVED** [review]
Scanned the code but didn't test the state machine carefully. I trust that your tests will cover most cases. Code coverage numbers would be interesting here though.

**Comment by ericstj** [comment]
Marking no-merge as I plan to move this to dotnet/runtime

**Comment by ericstj** [comment]
We're moving this code to dotnet/runtime

**Pull request closed without merging**